annotation/annotation/jobs/services.py

```diff
@@ -555,3 +581,139 @@ def update_jobs_names(db: Session, jobs_names: Dict):
     for key, value in jobs_names.items():
         db.query(Job).filter(Job.job_id == key).update({Job.name: value})
     db.commit()
+
+
+FileInWork = List[Dict[str, Union[int, List[int]]]]
+PagesInWork = Dict[str, FileInWork]
+
+
+def get_pages_in_work(
+    job_tasks: Set[ManualAnnotationTask],
+    tasks_to_delete: Set[ManualAnnotationTask],
+) -> PagesInWork:
+    task_type = {
+        True: "validation",
+        False: "annotation",
+    }
+    pages_in_work = {}
+    tasks_in_work = set(job_tasks) - tasks_to_delete
+
+    for task_in_work in tasks_in_work:
+        pages_in_work.setdefault(
+            task_type[task_in_work.is_validation], []
+        ).append(
+            {
+                "file_id": task_in_work.file_id,
+                "pages_number": task_in_work.pages,
+            }
+        )
+
+    return pages_in_work
+
+
+def get_tasks_to_delete(
+    job_tasks: List[ManualAnnotationTask],
+) -> Set[ManualAnnotationTask]:
+    started_annotation_tasks = [
+        task
+        for task in job_tasks
+        if (
+            task.status
+            in {
+                TaskStatusEnumSchema.in_progress,
+                TaskStatusEnumSchema.finished,
+            }
+            and not task.is_validation
+        )
+    ]
+    unstarted_tasks = [
+        task
+        for task in job_tasks
+        if task.status
+        in {TaskStatusEnumSchema.pending, TaskStatusEnumSchema.ready}
+    ]
+    annotation_tasks_to_delete = [
+        task for task in unstarted_tasks if not task.is_validation
+    ]
+
+    validation_tasks_to_delete: List[ManualAnnotationTask] = []
+    # select validation tasks without starting annotations
+    for validation_task in (
+        task for task in unstarted_tasks if task.is_validation
+    ):
+        if not any(
+            task
+            for task in started_annotation_tasks
+            if (
+                task.file_id == validation_task.file_id,
+                task.pages == validation_task.pages,
+            )
+        ):
+            validation_tasks_to_delete.append(validation_task)
+
+    return set(annotation_tasks_to_delete + validation_tasks_to_delete)
+
+
+Task = Dict[str, Union[int, List[int], bool]]
+
+
+def remove_pages_in_work(tasks: List[Task], pages_in_work: FileInWork):
+    """Remove already started pages from tasks. Removing inplace."""
+    tasks.sort(key=lambda x: x["pages"])
```

Comment:
Looks like the function does also an inplace sorting, which is not mentioned in function name and docstring